### PR TITLE
Improve documentation

### PR DESCRIPTION
**Pull request by alandefreitas**
This is really a draft. Locally, I still have some notes with questions to continue with this documentation. I didn't commit these but we should probably discuss them before moving on.

Some other commits in this PR fix small quickbook and doxygen problems in passing. We could cherry-pick some of these but there's nothing critical in there.

### CODE REVIEW

**Comment by vinniefalco** [comment]
Can we please rebase and squash this down?

**Comment by alandefreitas** [comment]
> Can we please rebase and squash this down?

I'm rebasing it but I wasn't planning to squash it down while this is still a draft. The reason is this PR has a lot of unrelated stuff and it was my understanding that you'd rather have that go into smaller PRs.

**Comment by alandefreitas** [comment]
So the PR has been rebased. The commits were squashed into fewer commits containing each of the unrelated changes in this PR:

- Include the C++11 requirement in the CMake target
- Fix lots of small quickbook and doxygen warnings
- Updated guidelines in javadoc.hpp
- Javadoc for `string_value`, `params_encoded`, and `params` (The javadoc is better than before but there are still some guidelines from javadoc.hpp missing)

If any of these changes are considered useful while this PR is still a draft, I can cherry-pick them into another branch and open another PR. If everything is considered useful, I can squash them into a single commit.

(This PR contains all these unrelated changes mostly because I wanted to take advantage of the common documentation preview while this is still a draft.)

File: CMakeLists.txt [hunk 1]
```diff
@@ -47,6 +47,7 @@ endif()
 
 function(boost_url_setup_properties target)
     target_compile_features(${target} PUBLIC cxx_constexpr)
+    target_compile_features(${target} PUBLIC cxx_std_11)
```
Comment (vinniefalco) [inline CMakeLists.txt:59]: Why are you fiddling with the CML?

**Comment by vinniefalco** [comment]
Changes to CMakeLists.txt should be in a separate pull request, in a single commit

File: doc/javadoc.hpp [hunk 2]
```diff
@@ -10,42 +10,63 @@
...
-    @see @ref {refid}, @ref {refid}.
-*/
+    @par References
+    @li <a href="">text</a>
+
+    @see
+        @ref {refid},
+        @ref {refid},
+        @ref {refid}.
+*/
```
Comment (vinniefalco) [inline doc/javadoc.hpp:72]: missing linefeed

File: doc/qbk/1.1.quicklook.qbk [hunk 3]
```diff
@@ -13,6 +13,8 @@ This section is intended to give the reader a brief overview of the features
 and interface style of the library.
 
 Say you have the following URL that you want to parse:
+
+[teletype]
```
Comment (vinniefalco) [inline doc/qbk/1.1.quicklook.qbk:17]: why `[teletype]`? This should not be needed

File: include/boost/url/authority_view.hpp [hunk 4]
```diff
@@ -1224,7 +1224,7 @@ BOOST_URL_DECL
 std::ostream&
 operator<<(
     std::ostream& os,
-    authority_view const& a);
+    authority_view const& u);
```
Comment (vinniefalco) [inline include/boost/url/authority_view.hpp:1227]: What's the point of renaming this? I was going to use `a` as the canonical variable name for authority views.

File: include/boost/url/impl/params.hpp [hunk 5]
```diff
@@ -285,6 +287,7 @@ assign(FwdIt first, FwdIt last,
 //
 //------------------------------------------------
 
+inline
```
Comment (vinniefalco) [inline include/boost/url/impl/params.hpp:290]: these were already marked inline in the first declaration, or else how do you think that this would link?

File: include/boost/url/params_view.hpp [hunk 6]
```diff
@@ -43,6 +43,18 @@ class params_view
 public:
     class iterator;
 
+    /** The value_type of query parameters
```
Comment (vinniefalco) [inline include/boost/url/params_view.hpp:46]: If we are going to accept this brief then it should say "The value type of query parameters"

File: include/boost/url/string.hpp [hunk 7]
```diff
@@ -20,33 +20,83 @@
...
-    buffer. They are reference counted which
-    makes copies cheap. The type is derived
-    from @ref string_view which provides
-    compatibility with strings in terms of
-    comparisons and converisons. However,
-    care must be exercised; undefined
+    buffer.
+
+    They are reference counted to
+    makes copies more efficient.
```
Comment (vinniefalco) [inline include/boost/url/string.hpp:30]: saying that copies are "more efficient" is worse than saying "cheap." More efficient could mean that copies are 10 times faster. Which is definitely more efficient, but is not "cheap" (consider copying a 10 megabyte string). "cheap" has a specific meaning in C++. For example, copying an iterator is supposed to be "cheap." Copying a string view is "cheap". etc.

File: include/boost/url/string.hpp [hunk 8]
```diff
@@ -20,33 +20,83 @@
...
+        This struct stores the number of references to this read-only string.
+        As the string is read-only, we can create a copy by simply updating
+        the reference counter.
+
+        As the underlying string is constructed with a custom allocator, the
+        struct also includes a custom destructor to deallocate the
+        string whenever the reference counter reaches zero.
+
+     */
     struct base;
```
Comment (vinniefalco) [inline include/boost/url/string.hpp:61]: Did you document a private nested type?

File: include/boost/url/string.hpp [hunk 9]
```diff
@@ -55,38 +105,115 @@ class string_value : public string_view
...
+
+     */
     class allocator;
 
-    inline
+    /** Destructor
+
+        This function decrements the string_value reference counter and
+        destroys the underlying string if no other instance of @ref string_value
+        refers to the same underlying string.
```
Comment (vinniefalco) [inline include/boost/url/string.hpp:125]: Upon destruction, ownership of the shared string is released. If there are no more owners, the resources for the string are released.

Comment (vinniefalco) [reply include/boost/url/string.hpp:125]: We usually never talk about reference counting, but instead discuss "shared ownership"

File: include/boost/url/string.hpp [hunk 10]
```diff
@@ -20,33 +20,83 @@
 namespace boost {
 namespace urls {
 
-/** A read-only, reference counted string
```
Comment (vinniefalco) [inline include/boost/url/string.hpp:23]: This should say "A const string with shared ownership"

File: CMakeLists.txt (hunk 1 above)
Comment (alandefreitas) [reply CMakeLists.txt:59]: The problem is because this is wrong or because it's unrelated to the PR?

Comment (vinniefalco) [reply CMakeLists.txt:59]: I have no idea, I am not a cmake expert, so I don't know what impact this has. And the other libraries don't have this line. So why don't they? The people who worked on the CMakeLists.txt decided not to include it. Why? Maybe it is not available in older cmakes. Maybe it causes a problem with the Boost superproject. Maybe they just didn't know it exists. In any event, this change is unrelated to improving documentation and it needs to be in its own pull request, reviewed by people who know Cmake better than I do (and who know the impact on the Boost superproject).

File: doc/javadoc.hpp [hunk 11]
```diff
@@ -10,42 +10,63 @@
 /** {brief}
 
     {description}
+    <!-- Assume the user knows C++ -->
+    <!-- Assume the user read the whole exposition -->
```
Comment (vinniefalco) [inline doc/javadoc.hpp:14]: read or "will read"

File: CMakeLists.txt (hunk 1 above)
Comment (alandefreitas) [reply CMakeLists.txt:59]: Assuming it's because it's unrelated to the PR, this is because this draft was moved from the fork because the PR in the fork didn't get the preview. The intention of the draft was to share the state of the documentation and so you could have a look at what I have locally to give me instructions. If anything is useful, I would just cherry-pick it into another PR. In any case, this doesn't seem to be working. I'll just close this PR then.

File: doc/qbk/1.1.quicklook.qbk (hunk 3 above)
Comment (alandefreitas) [reply doc/qbk/1.1.quicklook.qbk:17]: Without teletype, quickbook assumes this is c++. This generates a warning and highlights the syntax incorrectly.

File: include/boost/url/authority_view.hpp (hunk 4 above)
Comment (alandefreitas) [reply include/boost/url/authority_view.hpp:1227]: The var name wasn't matching the documentation. This generates a doxygen warning. So I'll change the documentation instead.

File: doc/qbk/1.1.quicklook.qbk (hunk 3 above)
Comment (vinniefalco) [reply doc/qbk/1.1.quicklook.qbk:17]: Ok so it is just for the non-c++ blocks. makes sense

File: include/boost/url/impl/params.hpp (hunk 5 above)
Comment (alandefreitas) [reply include/boost/url/impl/params.hpp:290]: I moved the keywords to the implementation. This was generating a warning in static analysis: https://isocpp.org/wiki/faq/inline-functions#where-to-put-inline-keyword

File: include/boost/url/string.hpp (hunk 7 above)
Comment (alandefreitas) [reply include/boost/url/string.hpp:30]: So should I say "cheap" or "constant time"?

File: include/boost/url/impl/params.hpp (hunk 5 above)
Comment (vinniefalco) [reply include/boost/url/impl/params.hpp:290]: hmm......... where is this warning? what does boost.json do?

File: doc/javadoc.hpp (hunk 11 above)
Comment (alandefreitas) [reply doc/javadoc.hpp:14]: I meant "read" in the past tense. Or a user who "has read" the whole exposition. 

I inferred this rule in the last PR, as you told me to assume in the reference API, the user has read the whole exposition.

File: include/boost/url/string.hpp (hunk 8 above)
Comment (vinniefalco) [reply include/boost/url/string.hpp:61]: We should never document private stuff, because the return on the investment is vanishingly small. First of all maintainers have a much better grasp of the library and its implementation than users. Second, users get zero value from documentation on private members. However, we should still use comments to explain non-obvious things in the implementation (inside function definitions). Almost always these comments should state "why" something is being done. Here's an example, we skip incrementing `it`:
https://github.com/vinniefalco/http_proto/blob/f2c3a5d9bfb49060af467b65843e125e6db958e2/include/boost/http_proto/rfc/impl/field_rule.ipp#L152

File: CMakeLists.txt (hunk 1 above)
Comment (alandefreitas) [reply CMakeLists.txt:59]: > this change is unrelated to improving documentation 

Since (i) this is a sufficient condition to remove the `cxx_std_11` requirement from this target and (ii) I'm closing this PR anyway, I won't argue in favor of it.

File: include/boost/url/impl/params.hpp (hunk 5 above)
Comment (alandefreitas) [reply include/boost/url/impl/params.hpp:290]: > hmm......... where is this warning? 

It's not a warning in our tooling. I don't even think we have common static analysis in our tooling. This is local. The tool points me to the isocpp guidelines in https://isocpp.org/wiki/faq/inline-functions#where-to-put-inline-keyword

The guideline seems reasonable unless there's another motivation I don't know about. But I can undo that if you want, of course. I'll just disable it locally so it stops annoying me.

File: include/boost/url/string.hpp (hunk 8 above)
Comment (alandefreitas) [reply include/boost/url/string.hpp:61]: > We should never document private stuff, because the return on the investment is vanishingly small

I wouldn't document everything that is private. I agree with the argument for not writing it. In this case, and only in this case, the marginal cost is zero because it was already done. So the return on investment in `NaN`.

> However, we should still use comments to explain non-obvious things in the implementation (inside function definitions). Almost always these comments should state "why" something is being done.

I can't disagree with that.

File: doc/qbk/1.1.quicklook.qbk (hunk 3 above)
Comment (alandefreitas) [reply doc/qbk/1.1.quicklook.qbk:17]: Yes. This command tells quickbook the next block is plain text.

Comment (alandefreitas) [reply doc/qbk/1.1.quicklook.qbk:17]: Quickbook chose the worst keyword possible for that here, but I have no power over that.

File: include/boost/url/impl/params.hpp (hunk 5 above)
Comment (vinniefalco) [reply include/boost/url/impl/params.hpp:290]: if we change it, we have to change it not only globally in url but also http_proto, json, and beast

File: doc/javadoc.hpp (hunk 11 above)
Comment (vinniefalco) [reply doc/javadoc.hpp]: well that's why I say "has read" _or_ "will read"

**Comment by vinniefalco** [comment]
Lets not put backticks in the commit message briefs because it makes them longer and less uniform

File: include/boost/url/impl/params.hpp (hunk 5 above)
Comment (alandefreitas) [reply include/boost/url/impl/params.hpp:290]: > if we change it, we have to change it not only globally in url but also http_proto, json, and beast

Well... that's not up to me then. I'll undo it and disable the warning locally.

File: doc/javadoc.hpp (hunk 11 above)
Comment (alandefreitas) [reply doc/javadoc.hpp]: Exactly. I've changed it to "has read". The present tense would be "reads" here, so it still wouldn't be ambiguous, but "has read" is better.

File: include/boost/url/string.hpp (hunk 7 above)
Comment (vinniefalco) [reply include/boost/url/string.hpp]: I like "cheap" because "constant time" sounds more precise, and I don't want to paint myself into a corner making complexity claims. Sometimes I do make complexity claims (for example in JSON) and in that case "constant time" is appropriate, but note that when you say "constant time" you are baking that into the public interface. If the algorithm changes to be no longer constant time, then that is technically breaking programs that depend on the complexity guarantee.

In short, "cheap" gives us some needed wiggle room and communicates the idea sufficiently, without locking us into making a guarantee.

**Comment by alandefreitas** [comment]
> Lets not put backticks in the commit message briefs because it makes them longer and less uniform

Done.

File: include/boost/url/string.hpp (hunk 7 above)
Comment (alandefreitas) [reply include/boost/url/string.hpp]: > "cheap" has a specific meaning in C++

> I like "cheap" because "constant time" sounds more precise

What's the specific meaning of "cheap" in C++ then? Less than O(log n)?

File: include/boost/url/impl/params.hpp (hunk 5 above)
Comment (vinniefalco) [reply include/boost/url/impl/params.hpp]: Lets open an issue for it, so we can track it

Comment (alandefreitas) [reply include/boost/url/impl/params.hpp]: It's already undone, but you can let me know if I should redo it at any time. 

My understanding is the isocpp guideline is correct and the reason Boost.URL does it seems to be because it's replicating a pattern from ASIO that doesn't apply here. To the best of my knowledge, Boost.URL doesn't have a macro intended to enable and disable these specifiers depending on how the ipp files are going to be used.

In any case, I don't think the issue is useful at this point because you have already stated you don't consider this to be an issue in this lib.

**Comment by alandefreitas** [comment]
I'm closing this since the draft preview is not very useful at the moment and everything else is either useless or wrong.

File: CMakeLists.txt (hunk 1 above)
Comment (vinniefalco) [reply CMakeLists.txt:59]: the c++11 change should be opened as its own separate pull request

File: doc/qbk/1.1.quicklook.qbk (hunk 3 above)
Comment (vinniefalco) [reply doc/qbk/1.1.quicklook.qbk:17]: we need this change

File: doc/qbk/2.8.authority_view.qbk [hunk 12]
```diff
@@ -25,6 +25,8 @@ The authority view provides the subset of observer member functions found in
 __url_view__ which are relevant to the authority. However, when an authority
 is parsed on its own, the leading double slashes ("//") are not present. The
 following authority string is valid for `parse_authority`:
+
+[teletype]
```
Comment (vinniefalco) [inline doc/qbk/2.8.authority_view.qbk:29]: need

File: include/boost/url/authority_view.hpp [hunk 13]
```diff
@@ -1218,7 +1218,7 @@ parse_authority(
 
     @param os The output stream to write to
 
-    @param u The URL to write
+    @param a The URL to write
```
Comment (vinniefalco) [inline include/boost/url/authority_view.hpp:1221]: need

File: include/boost/url/ipv6_address.hpp [hunk 14]
```diff
@@ -347,7 +347,7 @@ class ipv6_address
 
     @param os The output stream to write to.
 
-    @param addrr The address to write.
+    @param addr The address to write.
```
Comment (vinniefalco) [inline include/boost/url/ipv6_address.hpp:350]: need

File: include/boost/url/params.hpp [hunk 15]
```diff
@@ -106,104 +139,205 @@ class params
...
     using difference_type = std::ptrdiff_t;
 
-    //--------------------------------------------
-    //
-    // Members
-    //
-    //--------------------------------------------
+    /** <!-- @name Members -->
+        <!-- @{ -->
+     */
```
Comment (vinniefalco) [inline include/boost/url/params.hpp:169]: this grouping stuff doesn't work. please add this file back to the new PR

File: include/boost/url/params_encoded.hpp [hunk 16]
```diff
@@ -26,146 +26,312 @@ namespace urls {
 class url;
 #endif
 
+/** A random-access handle to percent-encoded query parameters in a url.
```
Comment (vinniefalco) [inline include/boost/url/params_encoded.hpp:29]: Add this file back to the new PR please

File: include/boost/url/params_view.hpp [hunk 17]
```diff
@@ -43,6 +43,18 @@ class params_view
...
+    /** The value type of query parameters
+
+        The query params object stores a list of encoded string view pairs
+        representing the query parameters identified in the original url.
+
+        When these values are dereference, they are lazily decoded into
+        constant read_only strings in params::reference.
+
+        A special boolean member `has_value` is used to differentiate
+        empty strings from non-existent strings.
```
Comment (vinniefalco) [inline include/boost/url/params_view.hpp:55]: Add this back

File: include/boost/url/string.hpp [hunk 18]
```diff
@@ -20,33 +20,82 @@
...
-    comparisons and converisons. However,
-    care must be exercised; undefined
+    buffer.
+
+    They are reference counted to allow cheap copies.
+
+    The type is derived from `string_view`,
+    which provides compatibility with strings
+    in terms of comparisons and conversions.
+
```
Comment (vinniefalco) [inline include/boost/url/string.hpp:34]: keep this change

File: include/boost/url/string.hpp [hunk 19]
```diff
@@ -20,33 +20,82 @@
...
     portion of the object is modified
     directly, for example by calling
     `remove_suffix` or `operator=`.
 
-    Slicing however, is supported, as
+    Slicing, however, is supported, as
     copies of the `string_view` portion
     of the object are valid and remain
-    valid for the lifetime of the oriignal
+    valid for the lifetime of the original
```
Comment (vinniefalco) [inline include/boost/url/string.hpp:44]: keep this

File: include/boost/url/string.hpp (hunk 8 above)
Comment (vinniefalco) [reply include/boost/url/string.hpp:60]: keep the file

File: include/boost/url/string.hpp [hunk 20]
```diff
@@ -55,38 +104,117 @@ class string_value : public string_view
...
+     */
     inline
     ~string_value();
 
+    /** Default constructor
+
+        This function creates the default base @ref string_view, which points to
+        no string / nullptr.
+
+        @note As an optimization, the reference counter is `nullptr`.
```
Comment (vinniefalco) [inline include/boost/url/string.hpp:134]: "@note As an optimization, the reference counter is `nullptr`." could be a comment in the implementation but it doesn't go in a javadoc

File: include/boost/url/string.hpp [hunk 21]
```diff
@@ -55,38 +104,117 @@ class string_value : public string_view
...
+    /** Construct an initial string_value
+
+        This function creates a string of the specified size at the
+        destination with the a custom allocator.
+
+        @note Although this function returns sets the destination pointer
+        to the new string address, this string_value owns the buffer.
+        The buffer is destroyed after the last string_value points to it.
+
+        @tparam Allocator Allocator type
```
Comment (vinniefalco) [inline include/boost/url/string.hpp:148]: you don't need to document the tparam for a well known concept. Try changing the declaration to say `__Allocator__` and see if docca turns it into a link.

File: include/boost/url/string.hpp [hunk 22]
```diff
@@ -55,38 +104,117 @@ class string_value : public string_view
...
+
+        @param n Number of chars in the string `dest`
+
+        @param a Allocator used to create a copy of `dest`
```
Comment (vinniefalco) [inline include/boost/url/string.hpp:152]: "The allocator to use"

File: include/boost/url/string_view.hpp [hunk 23]
```diff
@@ -32,7 +31,6 @@ using pos_t = std::size_t;
     owning the character buffer extends until
     the string view is no longer referenced.
 */
-//typedef boost::string_view string_view;
 typedef boost::core::string_view string_view;
```
Comment (vinniefalco) [inline include/boost/url/string_view.hpp:34]: yeah if you want this tidying, keep it
